File: mathsys/dev/typescript/level1.ts
//^
//^ HEAD
//^

//> HEAD -> DATA
import {TYPE, MODULE} from "./tokenizer";
import {NonTerminal} from "./nonterminal";
import {Opcode, Option, Group, Pointer, Binary, node, String, Vec} from "./ir";


//^
//^ 1ºLEVEL
//^

//> 1ºLEVEL -> NAMESPACE
export abstract class Level1 extends NonTerminal {}

//> 1ºLEVEL -> DECLARATION
export class Declaration extends Level1 {
    private code = new Opcode(0x02).binary();
    readonly group: string | null;
    readonly variable: Variable;
    readonly value: Level2;
    constructor(items: (TYPE | Variable | Level2)[]) {
        super();
        this.group = items.length === 3 ? (items[0] as TYPE).value : null;
        this.variable = (items.length === 3 ? items[1] : items[0]) as Variable;
        this.value = (items.length === 3 ? items[2] : items[1]) as Level2;
        this.freeze();
        return this;
    }
}

//> 1ºLEVEL -> TYPES
import {Start} from "./start";
import {Level2} from "./level2";
import {Variable} from "./level5";

File: dev/base/level1.rs
//^ 
//^ HEAD
//^

//> HEAD -> LOCAL
use super::nonterminal::{Spawn, NonTerminal};
use super::start::Start;


//^
//^ 1ºLEVEL
//^

//> 1ºLEVEL -> NAMESPACE
#[derive(Debug, Clone)]
pub enum Level1 {
    Declaration(Declaration),
    Definition(Definition),
    Annotation(Annotation),
    Node(Node),
    Equation(Equation),
    Use(Use)
}

//> 1ºLEVEL -> DECLARATION
#[derive(Debug, Clone)]
pub struct Declaration {} impl Spawn for Declaration {fn summon(items: Vec<NonTerminal>) -> NonTerminal {return NonTerminal::Level1(Level1::Declaration(Self {}))}}

//> 1ºLEVEL -> DEFINITION
#[derive(Debug, Clone)]
pub struct Definition {} impl Spawn for Definition {fn summon(items: Vec<NonTerminal>) -> NonTerminal {return NonTerminal::Level1(Level1::Definition(Self {}))}}

//> 1ºLEVEL -> ANNOTATION
#[derive(Debug, Clone)]
pub struct Annotation {} impl Spawn for Annotation {fn summon(items: Vec<NonTerminal>) -> NonTerminal {return NonTerminal::Level1(Level1::Annotation(Self {}))}}

//> 1ºLEVEL -> NODE
#[derive(Debug, Clone)]
pub struct Node {} impl Spawn for Node {fn summon(items: Vec<NonTerminal>) -> NonTerminal {return NonTerminal::Level1(Level1::Node(Self {}))}}

//> 1ºLEVEL -> EQUATION
#[derive(Debug, Clone)]
pub struct Equation {} impl Spawn for Equation {fn summon(items: Vec<NonTerminal>) -> NonTerminal {return NonTerminal::Level1(Level1::Equation(Self {}))}}

//> 1ºLEVEL -> USE
#[derive(Debug, Clone)]
pub struct Use {
    start: Option<Start>
} impl Spawn for Use {fn summon(items: Vec<NonTerminal>) -> NonTerminal {return NonTerminal::Level1(Level1::Use(Self {
    start: None
}))}}